pkg/microservice/warpdrive/core/service/taskplugin/job.go

```diff
@@ -272,13 +282,25 @@ func (b *JobCtxBuilder) BuildReaperContext(pipelineTask *task.Task, serviceName
 		ctx.Repos = append(ctx.Repos, repo)
 	}
 
+	envmaps := make(map[string]string)
 	for _, ev := range b.JobCtx.EnvVars {
 		val := fmt.Sprintf("%s=%s", ev.Key, ev.Value)
 		if ev.IsCredential {
 			ctx.SecretEnvs = append(ctx.SecretEnvs, val)
 		} else {
 			ctx.Envs = append(ctx.Envs, val)
 		}
+		envmaps[ev.Key] = ev.Value
+	}
+
+	if b.JobCtx.UploadEnabled {
+		for _, upload := range b.JobCtx.UploadInfo {
+			// since the frontend won't change the file path, the backend will have to add it
+			upload.FilePath = fmt.Sprintf("$WORKSPACE/%s", upload.FilePath)
+			// then we replace it with our env variables
+			upload.FilePath = replaceEnvWithValue(upload.FilePath, envmaps)
```

Comment:
What if the variables set by the user are not in the environment variables?

Reply:
It is not the user defined variables, it is ALL variables